paywall/src/data-iframe/blockchainHandler/purchaseKey.js

```diff
@@ -0,0 +1,168 @@
+import { getNetwork } from './network'
+import { getAccount } from './account'
+import ensureWalletReady from './ensureWalletReady'
+import { TRANSACTION_TYPES } from '../../constants'
+import pollForChanges from './pollForChanges'
+
+export async function purchaseKey(walletService, window, lock, amountToSend) {
+  await ensureWalletReady(window)
+  const account = getAccount()
+
+  return walletService.purchaseKey(lock, account, amountToSend)
+}
+
+/**
+ * Listens for the next time an event occurs, and resolves. If an error occurs first, it rejects
+ *
+ * @param {*} service either web3Service or walletService
+ * @param {string} event event to listen for
+ */
+function resolveOnEvent(service, event) {
+  let resolved = false
+  return new Promise((resolve, reject) => {
+    service.once('error', e => {
+      if (resolved) return
+      reject(e)
+    })
+    service.once(event, (...args) => resolve(args))
+  })
+}
+
+/**
+ * Does a basic shallow equal for transaction objects to see if any changes have occurred
+ * This is used to decide whether to trigger an update to the data
+ * @param {*} a old transaction object
+ * @param {*} b new transaction object
+ */
+function sortOfEqual(a, b) {
```

Comment:
Let's not write these things ourselves. I am sure there exist a dep which does this?

Reply:
I did research some deps. The primary option is `shallowEquals` but it's bulky. I'm going to go with `https://github.com/moroshko/shallow-equal`

Reply:
#3254